gemrb/core/PluginLoader.cpp

```diff
@@ -96,165 +93,130 @@ static inline voidvoid my_dlsym(void *handle, const char *symbol)
+	Log(MESSAGE, "PluginLoader", "Loaded plugin \"%s\" (%s).", desc.Description, pluginpath);
+	return true;
+}
 
+void LoadPlugins(const char* pluginpath)
+{
+	Log(MESSAGE, "PluginMgr", "Loading Plugins from %s", pluginpath);
 
+	DirectoryIterator dirIt(pluginpath);
+	if (!dirIt) {
+		Log(ERROR, "PluginMgr", "Plugin Directory (%s) does not exist!", pluginpath);
+		return;
+	}
 
-		// Try to load the Module
 #ifdef WIN32
-		HMODULE hMod = LoadLibrary( path );
+	const char* pluginExt = "dll";
 #else
-		// Note: the RTLD_GLOBAL is necessary to export symbols to modules
-		//       which python may have to dlopen (-wjp, 20060716)
-		// (to reproduce, try 'import bz2' or another .so module)
-		void* hMod = dlopen( path, RTLD_NOW | RTLD_GLOBAL );
+	const char* pluginExt = "so";
 #endif
-		if (hMod == NULL) {
-			Log(ERROR, "PluginLoader", "Cannot Load \"" TCHAR_FORMAT "\", skipping...", path);
-			continue;
-		}
 
-		//using C bindings, so we don't need to jump through extra hoops
-		//with the symbol name
-		Version_t LibVersion = ( Version_t ) (void*) GET_PLUGIN_SYMBOL( hMod, "GemRBPlugin_Version" );
-		Description_t Description = ( Description_t ) (void*) GET_PLUGIN_SYMBOL( hMod, "GemRBPlugin_Description" );
-		ID_t ID = ( ID_t ) (void*) GET_PLUGIN_SYMBOL( hMod, "GemRBPlugin_ID" );
-		Register_t Register = ( Register_t ) (void*) GET_PLUGIN_SYMBOL( hMod, "GemRBPlugin_Register" );
+	dirIt.SetFlags(DirectoryIterator::Files);
+	dirIt.SetFilterPredicate(new ExtFilter(pluginExt)); // rewinds
 
-		if (LibVersion==NULL) {
-			Log(ERROR, "PluginLoader", "Skipping invalid plugin \"" TCHAR_FORMAT "\".", path);
-			FREE_PLUGIN( hMod );
-			continue;
-		}
-		if (strcmp(LibVersion(), VERSION_GEMRB) ) {
-			Log(ERROR, "PluginLoader", "Skipping plugin \"" TCHAR_FORMAT "\" with version mistmatch.", path);
-			FREE_PLUGIN( hMod );
-			continue;
-		}
+	typedef std::set<std::string> PathSet;
+	PathSet delayedPlugins;
 
-		PluginDesc desc = { hMod, ID(), Description(), Register };
+	if (!dirIt) {
+		return;
+	}
 
-		if (libs.find(desc.ID) != libs.end()) {
-			Log(WARNING, "PluginLoader", "Plug-in \"" TCHAR_FORMAT "\" already loaded!", path);
-			FREE_PLUGIN( hMod );
+	char path[_MAX_PATH];
+	do {
+		const char *name = dirIt.GetName();
+		ieDword flags = 0;
+		core->plugin_flags->Lookup (name, flags);
+
+		// module is sent to the back
+		if (flags == PLF_DELAY) {
+			Log(MESSAGE, "PluginLoader", "Loading \"%s\" delayed.", name);
+			delayedPlugins.insert( name );
 			continue;
 		}
-		if (desc.Register != NULL) {
-			if (!desc.Register(PluginMgr::Get())) {
-				Log(WARNING, "PluginLoader", "Plugin Registration Failed! Perhaps a duplicate?");
-				FREE_PLUGIN( hMod );
-			}
+
+		// module is skipped
+		if (flags == PLF_SKIP) {
+			Log(MESSAGE, "PluginLoader", "Loading \"%s\" skipped.", name);
+			continue;
 		}
-		libs.insert(desc.ID);
 
-		Log(MESSAGE, "PluginLoader", "Loaded plugin \"%s\" (%s).", desc.Description, file);
+		PathJoin( path, pluginpath, name, NULL );
```

Comment:
Codacy found an issue: [Passing NULL after the last typed argument to a variadic function leads to undefined behaviour.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=6310305)